File: src-tauri/src/libs/traktor/nml_types.rs
// AIDEV-NOTE: Traktor NML type definitions
// Represents the structure of Traktor's collection.nml XML files
//
// Based on Traktor Pro 3.x NML format (VERSION="19")
// All fields are strings as they come from XML parsing
// Uses serde for XML deserialization with quick-xml

use serde::{Deserialize, Serialize};

/// Root NML structure
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorNML {
  #[serde(rename = "NML")]
  pub nml: NML,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NML {
  #[serde(rename = "@VERSION")]
  pub version: String,

  #[serde(rename = "HEAD")]
  pub head: TraktorHead,

  #[serde(rename = "COLLECTION")]
  pub collection: TraktorCollection,

  #[serde(rename = "PLAYLISTS")]
  pub playlists: Option<TraktorPlaylists>,

  #[serde(rename = "INDEXING")]
  pub indexing: Option<TraktorIndexing>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorHead {
  #[serde(rename = "@COMPANY")]
  pub company: String,

  #[serde(rename = "@PROGRAM")]
  pub program: String,
}

// ============================================================================
// Collection (Track Library)
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorCollection {
  #[serde(rename = "@ENTRIES")]
  pub entries: String,

  #[serde(rename = "ENTRY", default)]
  pub entry: Vec<TraktorEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorEntry {
  // Attributes on ENTRY element
  #[serde(rename = "@MODIFIED_DATE")]
  pub modified_date: Option<String>,

  #[serde(rename = "@MODIFIED_TIME")]
  pub modified_time: Option<String>,

  #[serde(rename = "@AUDIO_ID")]
  pub audio_id: Option<String>,

  #[serde(rename = "@TITLE")]
  pub title: Option<String>,

  #[serde(rename = "@ARTIST")]
  pub artist: Option<String>,

  // Child elements
  #[serde(rename = "LOCATION")]
  pub location: TraktorLocation,

  #[serde(rename = "ALBUM")]
  pub album: Option<TraktorAlbum>,

  #[serde(rename = "MODIFICATION_INFO")]
  pub modification_info: Option<TraktorModificationInfo>,

  #[serde(rename = "INFO")]
  pub info: Option<TraktorInfo>,

  #[serde(rename = "TEMPO")]
  pub tempo: Option<TraktorTempo>,

  #[serde(rename = "LOUDNESS")]
  pub loudness: Option<TraktorLoudness>,

  #[serde(rename = "MUSICAL_KEY")]
  pub musical_key: Option<TraktorMusicalKey>,

  #[serde(rename = "CUE_V2", default)]
  pub cue_v2: Vec<TraktorCue>,

  #[serde(rename = "PRIMARYKEY")]
  pub primarykey: Option<TraktorPrimaryKey>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorLocation {
  #[serde(rename = "@DIR")]
  pub dir: String,

  #[serde(rename = "@FILE")]
  pub file: String,

  #[serde(rename = "@VOLUME")]
  pub volume: Option<String>,

  #[serde(rename = "@VOLUMEID")]
  pub volumeid: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorAlbum {
  #[serde(rename = "@TITLE")]
  pub title: Option<String>,

  #[serde(rename = "@TRACK")]
  pub track: Option<String>,

  #[serde(rename = "@OF_TRACKS")]
  pub of_tracks: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorModificationInfo {
  #[serde(rename = "@AUTHOR_TYPE")]
  pub author_type: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorInfo {
  #[serde(rename = "@BITRATE")]
  pub bitrate: Option<String>,

  #[serde(rename = "@GENRE")]
  pub genre: Option<String>,

  #[serde(rename = "@LABEL")]
  pub label: Option<String>,

  #[serde(rename = "@COMMENT")]
  pub comment: Option<String>,

  #[serde(rename = "@COVERARTID")]
  pub coverartid: Option<String>,

  #[serde(rename = "@KEY")]
  pub key: Option<String>,

  #[serde(rename = "@PLAYTIME")]
  pub playtime: Option<String>,

  #[serde(rename = "@PLAYTIME_FLOAT")]
  pub playtime_float: Option<String>,

  #[serde(rename = "@RANKING")]
  pub ranking: Option<String>,

  #[serde(rename = "@IMPORT_DATE")]
  pub import_date: Option<String>,

  #[serde(rename = "@RELEASE_DATE")]
  pub release_date: Option<String>,

  #[serde(rename = "@LAST_PLAYED")]
  pub last_played: Option<String>,

  #[serde(rename = "@PLAYCOUNT")]
  pub playcount: Option<String>,

  #[serde(rename = "@FLAGS")]
  pub flags: Option<String>,

  #[serde(rename = "@FILESIZE")]
  pub filesize: Option<String>,

  #[serde(rename = "@COLOR")]
  pub color: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorTempo {
  #[serde(rename = "@BPM")]
  pub bpm: String,

  #[serde(rename = "@BPM_QUALITY")]
  pub bpm_quality: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorLoudness {
  #[serde(rename = "@PEAK_DB")]
  pub peak_db: Option<String>,

  #[serde(rename = "@PERCEIVED_DB")]
  pub perceived_db: Option<String>,

  #[serde(rename = "@ANALYZED_DB")]
  pub analyzed_db: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorMusicalKey {
  #[serde(rename = "@VALUE")]
  pub value: String,
}

/// Traktor Cue Point (V2 format)
///
/// Cue types:
/// - 0 = Cue (hot cue)
/// - 1 = Fade-in
/// - 2 = Fade-out
/// - 3 = Load
/// - 4 = Grid (beatgrid marker) - contains nested GRID element with BPM
/// - 5 = Loop
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorCue {
  #[serde(rename = "@NAME")]
  pub name: Option<String>,

  #[serde(rename = "@DISPL_ORDER")]
  pub displ_order: Option<String>,

  #[serde(rename = "@TYPE")]
  pub cue_type: String,

  #[serde(rename = "@START")]
  pub start: String,

  #[serde(rename = "@LEN")]
  pub len: Option<String>,

  #[serde(rename = "@REPEATS")]
  pub repeats: Option<String>,

  #[serde(rename = "@HOTCUE")]
  pub hotcue: Option<String>,

  #[serde(rename = "GRID")]
  pub grid: Option<TraktorGrid>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorGrid {
  #[serde(rename = "@BPM")]
  pub bpm: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorPrimaryKey {
  #[serde(rename = "@TYPE")]
  pub key_type: Option<String>,

  #[serde(rename = "@KEY")]
  pub key: Option<String>,
}

// ============================================================================
// Playlists Structure
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorPlaylists {
  #[serde(rename = "NODE")]
  pub node: TraktorNode,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorNode {
  #[serde(rename = "@TYPE")]
  pub node_type: String, // "FOLDER" or "PLAYLIST"

  #[serde(rename = "@NAME")]
  pub name: String,

  #[serde(rename = "SUBNODES")]
  pub subnodes: Option<TraktorSubnodes>,

  #[serde(rename = "PLAYLIST")]
  pub playlist: Option<TraktorPlaylistData>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorSubnodes {
  #[serde(rename = "@COUNT")]
  pub count: Option<String>,

  #[serde(rename = "NODE", default)]
  pub nodes: Vec<TraktorNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorPlaylistData {
  #[serde(rename = "@ENTRIES")]
  pub entries: String,

  #[serde(rename = "@TYPE")]
  pub playlist_type: String,

  #[serde(rename = "@UUID")]
  pub uuid: String,

  #[serde(rename = "ENTRY", default)]
  pub entry: Vec<TraktorPlaylistEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorPlaylistEntry {
  #[serde(rename = "PRIMARYKEY")]
  pub primarykey: TraktorPrimaryKey,
}

// ============================================================================
// INDEXING (Sorting Information)
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorIndexing {
  #[serde(rename = "SORTING_INFO", default)]
  pub sorting_info: Vec<TraktorSortingInfo>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorSortingInfo {
  #[serde(rename = "@PATH")]
  pub path: String,

  #[serde(rename = "CRITERIA")]
  pub criteria: Option<TraktorCriteria>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TraktorCriteria {
  #[serde(rename = "@ATTRIBUTE")]
  pub attribute: String,

  #[serde(rename = "@DIRECTION")]
  pub direction: String, // "UP" (ascending) or "DOWN" (descending)
}

File: old-electron/main/lib/traktor/types/nml-types.ts
/**
 * TypeScript interfaces for Traktor NML (Native Instruments Music Library) format.
 * Based on reverse-engineered XSD schema and real collection.nml data.
 *
 * These types match Traktor Pro 3.x NML format (VERSION="19").
 * All attributes are strings as they come from XML parsing.
 */

// ============================================================================
// Root NML Structure
// ============================================================================

export interface TraktorNML {
  NML: {
    VERSION: string;
    HEAD: TraktorHead;
    COLLECTION: TraktorCollection;
    PLAYLISTS?: TraktorPlaylists;
    SORTING_ORDER?: TraktorSortingOrder[];
    INDEXING?: TraktorIndexing; // Traktor uses this for sorting info, must be preserved
  };
}

/**
 * INDEXING section at the end of NML file.
 * Contains SORTING_INFO elements that Traktor uses for collection ordering.
 *
 * This section appears at the end of the NML file and must be preserved
 * for proper Traktor compatibility. Without it, Traktor may have issues reading the file.
 */
export interface TraktorIndexing {
  SORTING_INFO?: TraktorSortingInfo | TraktorSortingInfo[];
}

export interface TraktorSortingInfo {
  PATH: string; // e.g., "$COLLECTION", "Native Instruments"
  CRITERIA?: TraktorCriteria; // Optional sorting criteria
}

/**
 * Sorting criteria for a SORTING_INFO element.
 * Stores the column and direction used to sort a view.
 */
export interface TraktorCriteria {
  ATTRIBUTE: string; // Column to sort by: "TITLE", "ARTIST", "BPM", etc.
  DIRECTION: string; // "UP" (ascending) or "DOWN" (descending)
}

export interface TraktorHead {
  COMPANY: string;
  PROGRAM: string;
}

// ============================================================================
// Collection (Track Library)
// ============================================================================

export interface TraktorCollection {
  ENTRIES: string; // Number of entries as string
  ENTRY: TraktorEntry[];
}

export interface TraktorEntry {
  // Attributes on ENTRY element
  MODIFIED_DATE?: string; // Format: "2026/1/15"
  MODIFIED_TIME?: string; // Seconds since midnight as string
  AUDIO_ID?: string; // Base64 encoded audio fingerprint
  TITLE?: string;
  ARTIST?: string;

  // Child elements
  LOCATION: TraktorLocation;
  ALBUM?: TraktorAlbum;
  MODIFICATION_INFO?: TraktorModificationInfo;
  INFO?: TraktorInfo;
  TEMPO?: TraktorTempo;
  LOUDNESS?: TraktorLoudness;
  MUSICAL_KEY?: TraktorMusicalKey;
  CUE_V2?: TraktorCue | TraktorCue[]; // Can be single or array
  PRIMARYKEY?: TraktorPrimaryKey;
}

export interface TraktorLocation {
  DIR: string; // Format: "/:Users/:josev/:Music/:BOX/:2601/:"
  FILE: string; // Filename, may contain &amp; for &
  VOLUME: string; // Volume label, e.g., "C:"
  VOLUMEID?: string; // Volume ID hash
}

export interface TraktorAlbum {
  TITLE?: string;
  TRACK?: string; // Track number in album
  OF_TRACKS?: string; // Total tracks in album
}

export interface TraktorModificationInfo {
  AUTHOR_TYPE?: string; // e.g., "user"
}

export interface TraktorInfo {
  BITRATE?: string; // e.g., "320000" (bits per second)
  GENRE?: string;
  LABEL?: string; // Record label
  COMMENT?: string;
  COVERARTID?: string; // Cover art identifier
  KEY?: string; // Musical key in Traktor notation, e.g., "1m", "12d"
  PLAYTIME?: string; // Duration in seconds (integer)
  PLAYTIME_FLOAT?: string; // Duration in seconds (float)
  RANKING?: string; // 0-255 rating scale
  IMPORT_DATE?: string; // Format: "2026/1/15"
  RELEASE_DATE?: string; // Format: "2018/1/1"
  LAST_PLAYED?: string; // Format: "2026/1/15"
  PLAYCOUNT?: string; // Number of plays
  FLAGS?: string; // Internal flags
  FILESIZE?: string; // File size in KB
  COLOR?: string; // Track color in Traktor (1-16)
}

export interface TraktorTempo {
  BPM: string; // e.g., "123.000061"
  BPM_QUALITY?: string; // e.g., "100.000000"
}

export interface TraktorLoudness {
  PEAK_DB?: string; // e.g., "-0.973836"
  PERCEIVED_DB?: string; // e.g., "-0.627350"
  ANALYZED_DB?: string; // e.g., "-0.627350"
}

export interface TraktorMusicalKey {
  VALUE: string; // Numeric key value 0-23
}

/**
 * Traktor Cue Point (V2 format)
 *
 * Cue types:
 * - 0 = Cue (hot cue)
 * - 1 = Fade-in
 * - 2 = Fade-out
 * - 3 = Load
 * - 4 = Grid (beatgrid marker)
 * - 5 = Loop
 *
 * TYPE=4 (Grid/AutoGrid) cues contain a nested <GRID BPM="..."> element
 * that stores the precise BPM for beatgrid alignment. This must be preserved.
 */
export interface TraktorCue {
  NAME?: string; // e.g., "AutoGrid", "Intro"
  DISPL_ORDER?: string; // Display order
  TYPE: string; // 0-5, see above
  START: string; // Position in milliseconds
  LEN?: string; // Length in milliseconds (for loops)
  REPEATS?: string; // Loop repeats, -1 for infinite
  HOTCUE?: string; // Hotcue slot number 0-7
  GRID?: { BPM: string }; // Nested GRID element for TYPE=4 (AutoGrid) with precise BPM
}

export interface TraktorPrimaryKey {
  TYPE?: string;
  KEY?: string; // Usually the file path
}

// ============================================================================
// Playlists Structure
// ============================================================================

export interface TraktorPlaylists {
  NODE: TraktorNode;
}

/**
 * Recursive node structure for playlists and folders
 *
 * Traktor uses a tree structure where:
 * - TYPE="FOLDER" nodes contain other nodes
 * - TYPE="PLAYLIST" nodes contain track entries
 */
export interface TraktorNode {
  TYPE: 'FOLDER' | 'PLAYLIST';
  NAME: string;
  SUBNODES?: TraktorSubnodes;
  PLAYLIST?: TraktorPlaylistData;
}

export interface TraktorSubnodes {
  COUNT?: string;
  NODE?: TraktorNode | TraktorNode[]; // Can be single or array
}

export interface TraktorPlaylistData {
  ENTRIES: string; // Number of entries
  TYPE: string; // Playlist type
  UUID: string; // Unique identifier for sync
  ENTRY?: TraktorPlaylistEntry | TraktorPlaylistEntry[];
}

export interface TraktorPlaylistEntry {
  PRIMARYKEY: TraktorPrimaryKey;
}

// ============================================================================
// Sorting Order
// ============================================================================

export interface TraktorSortingOrder {
  PATH?: string;
  DIRECTION?: string;
}

// ============================================================================
// Constants
// ============================================================================

/**
 * Traktor cue type numeric values
 */
export enum TraktorCueType {
  CUE = 0,
  FADE_IN = 1,
  FADE_OUT = 2,
  LOAD = 3,
  GRID = 4,
  LOOP = 5,
}

/**
 * Traktor rating scale: 0-255 maps to 0-5 stars
 * Common values seen in collection.nml:
 * - 51 = 1 star
 * - 102 = 2 stars
 * - 153 = 3 stars
 * - 204 = 4 stars
 * - 255 = 5 stars
 */
export const TRAKTOR_RATING_SCALE = {
  MIN: 0,
  MAX: 255,
  STEP: 51, // 255 / 5 = 51
};

/**
 * Traktor key notation mapping
 *
 * Traktor uses Open Key notation:
 * - "1m" through "12m" for minor keys
 * - "1d" through "12d" for major keys (d = dur = major in German)
 */
export const TRAKTOR_KEY_MAP: Record<string, string> = {
  // Minor keys (m)
  '1m': 'Am',
  '2m': 'Em',
  '3m': 'Bm',
  '4m': 'F#m',
  '5m': 'Dbm',
  '6m': 'Abm',
  '7m': 'Ebm',
  '8m': 'Bbm',
  '9m': 'Fm',
  '10m': 'Cm',
  '11m': 'Gm',
  '12m': 'Dm',
  // Major keys (d = dur)
  '1d': 'C',
  '2d': 'G',
  '3d': 'D',
  '4d': 'A',
  '5d': 'E',
  '6d': 'B',
  '7d': 'Gb',
  '8d': 'Db',
  '9d': 'Ab',
  '10d': 'Eb',
  '11d': 'Bb',
  '12d': 'F',
};

/**
 * Reverse mapping from standard key notation to Traktor
 */
export const HARMONY_TO_TRAKTOR_KEY: Record<string, string> = Object.fromEntries(
  Object.entries(TRAKTOR_KEY_MAP).map(([traktor, harmony]) => [harmony, traktor]),
);

/**
 * MUSICAL_KEY VALUE to standard key mapping
 * Values 0-11 are major keys, 12-23 are minor keys
 */
export const TRAKTOR_MUSICAL_KEY_VALUES: Record<string, string> = {
  '0': 'C',
  '1': 'Db',
  '2': 'D',
  '3': 'Eb',
  '4': 'E',
  '5': 'F',
  '6': 'Gb',
  '7': 'G',
  '8': 'Ab',
  '9': 'A',
  '10': 'Bb',
  '11': 'B',
  '12': 'Cm',
  '13': 'Dbm',
  '14': 'Dm',
  '15': 'Ebm',
  '16': 'Em',
  '17': 'Fm',
  '18': 'Gbm',
  '19': 'Gm',
  '20': 'Abm',
  '21': 'Am',
  '22': 'Bbm',
  '23': 'Bm',
};
